taraxa/state/dpos/precompiled/dpos_contract.go

```diff
@@ -0,0 +1,1022 @@
+
+var ErrInsufficientBalance = util.ErrorString("Insufficient balance")
+var ErrNonExistentValidator = util.ErrorString("Validator does not exist")
+var ErrNonExistentDelegation = util.ErrorString("Delegation does not exist")
+var ErrExistentUndelegation = util.ErrorString("Undelegation already exist")
+var ErrNonExistentUndelegation = util.ErrorString("Undelegation does not exist")
+var ErrNonReadyUndelegation = util.ErrorString("Undelegation is not yet ready to be withdrawn")
+var ErrExistentValidator = util.ErrorString("Validator already exist")
+var ErrBrokenState = util.ErrorString("Fatal error state is broken")
+var ErrValidatorsMaxStakeExceeded = util.ErrorString("Validator's max stake exceeded")
+var ErrInsufficientDelegation = util.ErrorString("Insufficient delegation")
+
+var ErrTransferAmountIsZero = util.ErrorString("transfer amount is zero")
+var ErrWithdrawalExceedsDeposit = util.ErrorString("withdrawal exceeds prior deposit value")
+var ErrInsufficientBalanceForDeposits = util.ErrorString("insufficient balance for the deposits")
+var ErrCallIsNotToplevel = util.ErrorString("only top-level calls are allowed")
+var ErrNoTransfers = util.ErrorString("no transfers")
+var ErrDuplicateBeneficiary = util.ErrorString("duplicate beneficiary")
+
+// Contract storage fields keys
+var (
+	field_validators    = []byte{0}
+	field_state         = []byte{1}
+	field_delegations   = []byte{2}
+	field_undelegations = []byte{3}
+
+	field_eligible_count      = []byte{4}
+	field_eligible_vote_count = []byte{5}
+	field_amount_delegated    = []byte{6}
+)
+
+var Big10000 = new(big.Int).SetInt64(10000)
+
+// Maximum number of validators per batch returned by getValidators call
+const GetValidatorsMaxCount = 50
+
+// Maximum number of validators per batch returned by getDelegatorDelegations call
+const GetDelegatorDelegationsMaxCount = 50
+
+type State struct {
+	RwardsPer1Stake *big.Int
+
+	// number of references
+	Count uint32
+}
+
+type Contract struct {
+	cfg            Config
+	storage        StorageWrapper
+	delayedStorage Reader
+	Abi            abi.ABI
+
+	validators    Validators
+	delegations   Delegations
+	undelegations Undelegations
+
+	eligible_count_orig      uint64
+	eligible_count           uint64
+	eligible_vote_count_orig uint64
+	eligible_vote_count      uint64
+	amount_delegated_orig    *big.Int
+	amount_delegated         *big.Int
+	lazy_init_done           bool
+}
+
+func (self *Contract) Init(cfg Config, storage Storage, readStorage Reader) *Contract {
+	self.cfg = cfg
+	self.storage.Init(storage)
+	self.delayedStorage = readStorage
+	self.Abi, _ = abi.JSON(strings.NewReader(TaraxaDposClientMetaData))
+
+	self.validators.Init(&self.storage, field_validators)
+	self.delegations.Init(&self.storage, field_delegations)
+	self.undelegations.Init(&self.storage, field_undelegations)
+
+	return self
+}
+
+func (self *Contract) UpdateStorage(readStorage Reader) {
+	self.delayedStorage = readStorage
+}
+
+///// DO WE NEED THIS ? ////
```

Comment:
That was needed for mainnet hardfork. Also could be useful for future hardforks when we would need to change configuration related to this contract.

Reply:
@MatusKysel if you are keeping it, remove DO WE NEED THIS ? comment